### ISSUE TITLE
Missing gradients when transforming rendered mesh

### ISSUE THREAD

**Issue by olokobayusuf** [facebookresearch/pytorch3d#1386]
I am trying to use PyTorch3D as a differentiable renderer for a spatial reasoning task. While the examples provide a way to optimize the camera pose, there isn't a corresponding example for optimizing meshes that are rendered. And while [this issue](https://github.com/facebookresearch/pytorch3d/issues/1087) briefly touches on the topic, there is no confirmed resolution. I cooked up a small repro:
```py
from PIL import Image
from pytorch3d.renderer import BlendParams, FoVPerspectiveCameras, HardFlatShader, MeshRasterizer, MeshRenderer, RasterizationSettings
from pytorch3d.structures import Meshes
from pytorch3d.transforms import Transform3d, Translate, RotateAxisAngle
from pytorch3d.utils import checkerboard
from torch import tensor, zeros
from torch.nn import L1Loss
from torch.optim import SGD
from torchvision.transforms import ToTensor

# Load image
image = Image.open("test/media/2_floor.jpg")

# Create camera
R = tensor([
    [-0.81158288, -0.01076309, -0.58413816  ],
    [-0.01360212,  0.99990737,  0.00047447  ],
    [ 0.58407895,  0.00833059, -0.8116541   ],
])
T = tensor([0., -2., 5.])
cameras = FoVPerspectiveCameras(fov=96.2373816791421, degrees=True, R=R[None,...], T=T[None,...])

# Create mesh renderer
blend_params = BlendParams(background_color=(0., 0., 0.))
raster_settings = RasterizationSettings(image_size=(image.height, image.width), faces_per_pixel=1)
rasterizer = MeshRasterizer(cameras, raster_settings)
shader = HardFlatShader(cameras=cameras, blend_params=blend_params)
renderer = MeshRenderer(rasterizer, shader)

# Define mesh transform
def transform_mesh (mesh: Meshes, transform: Transform3d):
    updated_verts = transform.transform_points(mesh.verts_padded())
    result = mesh.update_padded(updated_verts)
    return result

# Create default plane
plane = checkerboard(1, (1, 1, 1), (1, 1, 1))

# Create optimizer
plane_rotation = RotateAxisAngle(90., axis="X")
plane_position = zeros(1, 3, requires_grad=True)
optimizer = SGD([plane_position], lr=1., momentum=0.)

# Create loss
l1_loss = L1Loss()
target = ToTensor()(image).unsqueeze(dim=0).permute(0, 2, 3, 1) # (N,H,W,C)

# Optimize
for step in range(10):
    # Forward pass
    plane_translation = Translate(plane_position)
    transform = plane_rotation.compose(plane_translation)
    world_plane = transform_mesh(plane, transform)
    plane_render = renderer(world_plane)
    # Retain gradients
    world_plane.verts_packed().retain_grad()
    plane_render.retain_grad()
    # Backprop
    optimizer.zero_grad()
    loss = l1_loss(target, plane_render[...,:3])
    loss.backward()
    #optimizer.step()
    # Check gradients
    print("Rendering:", plane_render.grad.count_nonzero())
    print("Vertices:", world_plane.verts_packed().grad.count_nonzero())
    break
```
Which uses this image as `test/media/2_floor.jpg`:

![2_floor](https://user-images.githubusercontent.com/17663473/203129363-b9c12342-fc08-44b6-9060-e8450f793a24.jpg)

The result is the following:
```
Rendering: tensor(504216)
Vertices: tensor(0)
```
As you can see, the gradients of `Meshes.verts_packed()` is all-zero, which means we can't backprop through any transformations on the vertices of a mesh. How can this issue be addressed or fixed?

**Comment by bottler** [comment]
I recommend constructing the Meshes explicitly (from faces and verts which come out of an original checkerboard and with a transform applied), avoiding update_padded, which makes it easy to see exactly what your calculation graph is meant to be. In a simple gradient descent like this, there should be no need to call retain_grad. The model calculation should be very clear.

You are trying to reposition the mesh based on matching up black and white areas. This may be difficult in general, but it will be impossible with HardFlatShader, because there is no smooth transition between black and white. Think about where you want the gradients to flow between `plane_position` and `loss`. Perhaps adjust the model until you verify that loss changes when position changes very slightly.

**Comment by olokobayusuf** [comment]
@bottler I haven't been able to follow up on this, but your last point is a most salient one:

> You are trying to reposition the mesh based on matching up black and white areas. This may be difficult in general, but it will be impossible with HardFlatShader, because there is no smooth transition between black and white.

And I got the clue that this could be the reason from working on a totally different problem: backpropagating through `torch.nn.functional.interpolate` with `mode="nearest"`. From PyTorch 3D's API, it seems that `BlendParams` *should* be a helpful construct for blending gradients across frame buffer texels? The docs are very sparse. Otherwise I guess I have to use a soft shader?

**Comment by bottler** [comment]
BlendParams contains parameters used by other objects, its presence doesn't indicate what the shader does. The user needs to choose a shader (you can see all the in-built ones [here](https://github.com/facebookresearch/pytorch3d/blob/main/pytorch3d/renderer/mesh/shader.py)) based on what they want to achieve. SoftPhongShader and SoftGouraudShader are likely wanted here.

**Comment by olokobayusuf** [comment]
I'm closing this issue, as @bottler's suggestion on using a soft shader is correct. Thanks!
